File: daily/2024/494_Target_Sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        if (n == 0) return 0;
        unordered_map<int,int> m;
        m[nums[0]] += 1;
        m[-nums[0]] += 1;
        for (int i = 1; i < n; i++){
            int curr = nums[i];
            unordered_map<int,int> temp;
            for (auto k : m){
                int val = k.first;
                int count = k.second;
                temp[val+curr] += count;
                temp[val-curr] += count;
            }
            swap(m, temp);
        }
        return m[target];
    }
};

int main(){
    vector<int> nums = {1,1,1,1,1};
    int target = 3;
    // vector<int> nums = {1};
    // int target = 1;
    // vector<int> nums = {0,0,0,0,0,0,0,0,1};
    // int target = 1;

    Solution S;
    cout << S.findTargetSumWays(nums, target) << endl;
    return 0;
}

File: daily_python/494_Target_Sum.py
import os
import sys
from typing import List
from collections import defaultdict

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = defaultdict(int)
        dp[nums[0]] += 1
        dp[-nums[0]] += 1
        for i in range(1, len(nums)):
            curr = nums[i]
            next_dp = defaultdict(int)
            for val, count in dp.items():
                next_dp[val+curr] += count
                next_dp[val-curr] += count
            dp = next_dp
        return dp[target]
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            target = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((nums, target))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, target in dataset:
        results.append(solution.findTargetSumWays(nums, target))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')